### PR TITLE
Remove singleton classes

### PR DESCRIPTION
**Pull request by PoojaBabusing**
## What?
This PR Consist of changes  that changes related to removal of  singleton class.
## Why?
Issues with singleton is anyone can access every time and there are multiple callers and this  may break the flow in long term, it can't be more scalable and might affect stability.
## How?
Removed all singleton classes.
## Testing?
Tested 7 Transaction and the speed is Good. 
## Screenshots (optional)
N/A
## Anything Else?
But some issue in develop causing disconnect in android on 7th attempt.

**Issue Link :**  https://github.com/mosip/inji/issues/546

### CODE REVIEW

File: ios/Wallet/Wallet+Extension.swift [hunk 1]
```diff
@@ -0,0 +1,55 @@
+//
```
Comment (shreeThaanuTw) [inline ios/Wallet/Wallet+Extension.swift]: Remove the code comments Header

File: ios/ble/Utility/chunker.swift [hunk 2]
```diff
@@ -37,7 +37,7 @@ class Chunker {
 
     func getTotalChunkCount(dataSize: Int) -> Double {
         var resulydouble = Double(dataSize)/Double(effectivePayloadSize)
-        return Double(ceill(resulydouble))
+        return Double(ceill(Float80(resulydouble)))
```
Comment (shreeThaanuTw) [inline ios/ble/Utility/chunker.swift]: change the variable name to something meaningful like **totalChukCount**

File: ios/Wallet/Wallet+Extension.swift [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+    func onDisconnectStatusChange(data: Data?) {
+        print("Handling notification for disconnect handle")
+        if let data {
+            let connStatusID = Int(data[0])
+            if connStatusID == 1 {
+                print("con statusid:", connStatusID)
+                destroyConnection()
+            }
+        } else {
+            print("weird reason!!")
```
Comment (shreeThaanuTw) [inline ios/Wallet/Wallet+Extension.swift]: remove the unwanted print comments, add more meaningful comments

**Review by shreeThaanuTw — COMMENTED** [review]
Can you please fix these things and for review

File: ios/Openid4vpBle/Openid4vpBle.swift [hunk 4]
```diff
@@ -22,10 +22,15 @@ class Openid4vpBle: RCTEventEmitter {
 
     @objc(setConnectionParameters:)
     func setConnectionParameters(params: String) -> Any {
-        var paramsObj = stringToJson(jsonText: params)
-        var firstPartOfPk = paramsObj["pk"]
-        //os_log(.info, "synchronized setConnectionParameters called with %{public}s and %{public}s ", params, firstPartOfPk)
-        Wallet.shared.setAdvIdentifier(identifier: firstPartOfPk as! String)
+        print("SetConnectionParameters->Params::\(params)")
+        let paramsObj = stringToJson(jsonText: params)
+        let firstPartOfPk = paramsObj["pk"] as? String
```
Comment (krishnakumar4a4) [inline ios/Openid4vpBle/Openid4vpBle.swift]: this variable should be renamed

File: ios/Openid4vpBle/Openid4vpBle.swift [hunk 5]
```diff
@@ -22,10 +22,15 @@ class Openid4vpBle: RCTEventEmitter {
...
+        print("synchronized setConnectionParameters called with", params, "and", firstPartOfPk)
+        wallet = Wallet()
+        wallet?.setTuvaliVersion(tuvaliVersion)
```
Comment (krishnakumar4a4) [inline ios/Openid4vpBle/Openid4vpBle.swift]: Don't expose `setTuvaliVersion`, instead read it from Utility class (similar to Android)

File: ios/ble/Utility/TransferHandler.swift [hunk 6]
```diff
@@ -67,15 +67,17 @@ class TransferHandler {
...
+            }
+            // checks if no more missing chunks exist on verifier
         }
         sendMessage(message: imessage(msgType: .READ_TRANSMISSION_REPORT, data: nil))
     }
 
     private func requestTransmissionReport() {
         var notifyObj: Data
-        Central.shared.write(serviceUuid: BLEConstants.SERVICE_UUID, charUUID: NetworkCharNums.TRANSFER_REPORT_REQUEST_CHAR_UUID, data: withUnsafeBytes(of: 1.littleEndian) { Data($0) })
+        delegate?.write(serviceUuid: BLEConstants.SERVICE_UUID, charUUID: NetworkCharNums.TRANSFER_REPORT_REQUEST_CHAR_UUID, data: withUnsafeBytes(of: 1.littleEndian) { Data($0) }, withResponse: false)
```
Comment (krishnakumar4a4) [inline ios/ble/Utility/TransferHandler.swift]: This was changed recently to write with response in the last release and there is retry mechanism on top of this.

File: ios/ble/central/Central.swift [hunk 7]
```diff
@@ -70,4 +68,13 @@ class Central: NSObject, CBCentralManagerDelegate {
             connectedPeripheral.writeValue(messageData, for: characteristic, type: .withoutResponse)
         }
     }
+    
+    func onDeviceDisconnected(isManualDisconnect: Bool) {
```
Comment (krishnakumar4a4) [inline ios/ble/central/Central.swift]: This exact function already exists on Wallet.swift file, why is it needed here again?
Also, the function signature seems to be different from Android implementation.

File: ios/Wallet/Wallet.swift [hunk 8]
```diff
@@ -29,9 +30,17 @@ class Wallet: NSObject {
...
+        //central?.tuvaliVersion = version
+    }
+    
+    func startScanning(){
+        central?.walletDelegate = self
+    }
 
     func destroyConnection(isSelfDisconnect: Bool){
-        onDeviceDisconnected(isSelfDisconnect: isSelfDisconnect)
+        central?.onDeviceDisconnected(isManualDisconnect: isSelfDisconnect)
```
Comment (krishnakumar4a4) [inline ios/Wallet/Wallet.swift]: I think we should separate out function that does disconnect Vs function which handles disconnect event.

Here destroyConnection calling onDeviceDisconnected is not making sense.
Can you refer naming conventions from Android side?

File: ios/ble/central/CentralManagerDelegate.swift [hunk 9]
```diff
@@ -12,8 +12,8 @@ extension Central {
             let scanResponseData = dataDict?[CBUUID(string: "AB2A")]  as! Data
             let advertisementData = dataDict?[CBUUID(string: "AB29")]  as! Data
             let publicKeyData =  advertisementData.subdata(in: advertisementData.count-5..<advertisementData.count) + scanResponseData
-            if Wallet.shared.isSameAdvIdentifier(advertisementPayload: advertisementData) {
-                Wallet.shared.setVerifierPublicKey(publicKeyData: publicKeyData)
+            print("veri pub key::", publicKeyData)
+            walletDelegate?.hasSameIdentifier(payload: advertisementData, publicData: publicKeyData) {
```
Comment (krishnakumar4a4) [inline ios/ble/central/CentralManagerDelegate.swift]: This line is missed out here

`Wallet.shared.setVerifierPublicKey(publicKeyData: publicKeyData)`

File: ios/ble/Utility/TransferHandler.swift [hunk 10]
```diff
@@ -99,8 +101,10 @@ class TransferHandler {
 
     private func sendResponseSize(size: Int) {
         let decimalString = String(size)
-        let d = decimalString.data(using: .utf8)
-        Central.shared.write(serviceUuid: Peripheral.SERVICE_UUID, charUUID: NetworkCharNums.RESPONSE_SIZE_CHAR_UUID, data: d!)
+        if let data = decimalString.data(using: .utf8) {
+            print(data)
```
Comment (krishnakumar4a4) [inline ios/ble/Utility/TransferHandler.swift]: change print statements to os_log statements

**Review by krishnakumar4a4 — CHANGES_REQUESTED** [review]
Please address the comments.

File: ios/ble/Utility/TransferHandler.swift (hunk 10 above)
Comment (shreethaanu) [reply ios/ble/Utility/TransferHandler.swift]: this log is for developer purpose, I'll remove this.

File: ios/ble/central/CentralManagerDelegate.swift (hunk 9 above)
Comment (shreeThaanuTw) [reply ios/ble/central/CentralManagerDelegate.swift]: Moved the condition check also inside the delegate method for more cleaner approach.

File: ios/Openid4vpBle/Openid4vpBle.swift (hunk 4 above)
Comment (shreeThaanuTw) [reply ios/Openid4vpBle/Openid4vpBle.swift]: Changed to privateKey & connectionParameter

File: ios/ble/Utility/TransferHandler.swift (hunk 6 above)
Comment (shreeThaanuTw) [reply ios/ble/Utility/TransferHandler.swift]: We have added a unique write method which takes in with or without response as a parameter and change based on that.

File: ios/ble/central/Central.swift (hunk 7 above)
Comment (shreeThaanuTw) [reply ios/ble/central/Central.swift]: Removed the function in wallet. added because of merge

File: ios/Openid4vpBle/Openid4vpBle.swift (hunk 5 above)
Comment (shreeThaanuTw) [reply ios/Openid4vpBle/Openid4vpBle.swift]: NA in the PR. So removed the function.

File: ios/Openid4vpBle/Openid4vpBle.swift [hunk 11]
```diff
@@ -48,16 +52,14 @@ class Openid4vpBle: RCTEventEmitter {
...
+        tuvaliVersion = version
+        os_log("Tuvali version - %{public}@",tuvaliVersion);
         return tuvaliVersion
       }
 
     @objc(destroyConnection:)
     func destroyConnection(withCallback callback: @escaping RCTResponseSenderBlock) -> Any {
-        Wallet.shared.destroyConnection(isSelfDisconnect: true)
-        callback([])
+        wallet?.handleDestroyConnection(isSelfDisconnect: true)
```
Comment (krishnakumar4a4) [inline ios/Openid4vpBle/Openid4vpBle.swift:62]: Once the destroy is done, I think we should reset the wallet variable to `nil` (consistent with Android impl)

File: ios/ble/central/Central.swift [hunk 12]
```diff
@@ -70,4 +68,14 @@ class Central: NSObject, CBCentralManagerDelegate {
...
         }
     }
+    
+    func disconnectAndClose() {
+        if let connectedPeripheral = self.connectedPeripheral {
+            centralManager.cancelPeripheralConnection(connectedPeripheral)
+        }
+    }
+
+    func destroyConnection(){
```
Comment (krishnakumar4a4) [inline ios/ble/central/Central.swift]: 1. Move this method into Wallet.swift file. The method scope should not leak into Central layer. If required place a delegate method as part of Wallet/PeripheralCommunicatorDelegate to call this method from lower layers.
2. Also rename this function to `onDeviceDisconnected` for consistency

File: ios/ble/peripheral/PeripheralCommunicatorDelegate.swift [hunk 13]
```diff
@@ -11,4 +11,6 @@ protocol PeripheralCommunicatorProtocol: AnyObject {
 protocol WalletProtocol: AnyObject {
     func onIdentifyWriteSuccess()
     func onDisconnectStatusChange(data: Data?)
+    func createConnectionHandler()
```
Comment (krishnakumar4a4) [inline ios/ble/peripheral/PeripheralCommunicatorDelegate.swift:14]: On what basis are we putting a method in either WalletProtocol or PeripheralCommunicatorProtocol?

For example:
`createConnectionHandler` is being called from `PeripheralDelegate` while `setVeriferKeyOnSameIdentifier` is called from `CentralManagerDelegate`

File: ios/ble/Utility/TransferHandler.swift [hunk 14]
```diff
@@ -186,7 +190,7 @@ extension TransferHandler: PeripheralCommunicatorProtocol {
             } else if status == 1 {
                 EventEmitter.sharedInstance.emitNearbyMessage(event: "send-vc:response", data: "\"REJECTED\"")
             }
-            Wallet.shared.destroyConnection(isSelfDisconnect: true)
+            destroyConnection?()
```
Comment (krishnakumar4a4) [inline ios/ble/Utility/TransferHandler.swift:193]: I think this should be calling `handleDestroyConnection` method, not `destroyConnection` method. Cross check with Android implementation in `onNotificationReceived` method in `Wallet.kt` file.

Also this is the confusion we get into if we don't differentiate method names based on what it does. From method names `handleDestroyConnection` and `destroyConnection` sounds same. But both are doing opposite activities as per the current implementation.

File: ios/ble/central/Central.swift [hunk 15]
```diff
@@ -70,4 +68,14 @@ class Central: NSObject, CBCentralManagerDelegate {
             connectedPeripheral.writeValue(messageData, for: characteristic, type: .withoutResponse)
         }
     }
+    
+    func disconnectAndClose() {
```
Comment (krishnakumar4a4) [inline ios/ble/central/Central.swift]: call it just `disconnect` if it is not doing Close, unlike Android.

**Comment by krishnakumar4a4** [comment]
@shreeThaanuTw and @dhivya17 Please address the comments.

File: ios/ble/peripheral/PeripheralCommunicatorDelegate.swift (hunk 13 above)
Comment (dhivya0413) [reply ios/ble/peripheral/PeripheralCommunicatorDelegate.swift:14]: WalletProtocol and PeripheralCommunicatorProtocol methods are created to communicate with Wallet and Transferhandler(peripheral stuff) from Central respectively.

File: ios/ble/Utility/TransferHandler.swift (hunk 14 above)
Comment (dhivya0413) [reply ios/ble/Utility/TransferHandler.swift:193]: This destroyConnection is a callback from transferHandler to wallet which actually invokes wallet handleDestroyConnection.

File: ios/ble/central/Central.swift (hunk 15 above)
Comment (dhivya0413) [reply ios/ble/central/Central.swift]: done with suggested changes.

File: ios/ble/central/Central.swift (hunk 12 above)
Comment (dhivya0413) [reply ios/ble/central/Central.swift]: done with suggested changes.

File: ios/Openid4vpBle/Openid4vpBle.swift (hunk 11 above)
Comment (dhivya0413) [reply ios/Openid4vpBle/Openid4vpBle.swift:62]: done with suggested changes.

**Comment by shreeThaanuTw** [comment]
I have tested the send VC, Disconnect and reject scenario.

File: ios/Openid4vpBle/Openid4vpBle.swift [hunk 16]
```diff
@@ -22,10 +22,14 @@ class Openid4vpBle: RCTEventEmitter {
 
     @objc(setConnectionParameters:)
     func setConnectionParameters(params: String) -> Any {
-        var paramsObj = stringToJson(jsonText: params)
-        var firstPartOfPk = paramsObj["pk"]
-        //os_log(.info, "synchronized setConnectionParameters called with %{public}s and %{public}s ", params, firstPartOfPk)
-        Wallet.shared.setAdvIdentifier(identifier: firstPartOfPk as! String)
+        print("SetConnectionParameters->Params::\(params)")
+        let connectionParameter = stringToJson(jsonText: params)
+        let privateKey = connectionParameter["pk"] as? String
```
Comment (krishnakumar4a4) [inline ios/Openid4vpBle/Openid4vpBle.swift:27]: This one is part of public key data, not private key

please change it to `advPayload`

Comment (krishnakumar4a4) [reply ios/Openid4vpBle/Openid4vpBle.swift:27]: @dhivya0413 @shreeThaanuTw Please make this change on develop

File: ios/ble/Utility/TransferHandler.swift (hunk 14 above)
Comment (krishnakumar4a4) [reply ios/ble/Utility/TransferHandler.swift:193]: Okay.

File: ios/Openid4vpBle/Openid4vpBle.swift (hunk 16 above)
Comment (shreethaanu) [reply ios/Openid4vpBle/Openid4vpBle.swift:27]: Ok, I'll make the change

**Pull request merged**